### ISSUE TITLE
Sauce Connect: Connection is shut down prematurely

### ISSUE THREAD

**Issue by bobholt** [travis-ci/travis-ci#7611]
This is an intermittent issue. About 50% of the time, before running tests, the Sauce Connect VM is shut down. Presumably this is a process-end signal from Travis:

```
13 Apr 12:59:15 - Remote tunnel VM has been terminated by user request. Shutting down.
13 Apr 12:59:15 - Cleaning up.
13 Apr 12:59:15 - Removing tunnel 88b7408f31994425bcbfa652d8320ae6.
13 Apr 12:59:15 - Waiting for any active jobs using this tunnel to finish.
13 Apr 12:59:15 - Press CTRL-C again to shut down immediately.
13 Apr 12:59:15 - Note: if you do this, tests that are still running will fail.
13 Apr 12:59:16 - All jobs using the tunnel have finished.
13 Apr 12:59:16 - Waiting for the connection to terminate...
13 Apr 12:59:17 - Connection closed (8).
13 Apr 12:59:17 - Goodbye.
```

Re-running the job once or twice hits the scenario where the connection stays open and the build is able to complete correctly.

When I run Sauce Connect via my local machine, I have a 100% success rate running the test suite. The VM does not shut down prematurely. This contributes to my belief that this is a Travis issue and not a Sauce issue.

See https://travis-ci.org/bobholt/web-platform-tests/jobs/221719688#L478-L487

**Comment by BanzaiMan** [comment]
Your build has multiple jobs, all of which initiates Sauce Connect connections. Is it possible that you are running into limits because of this? When you try it locally, are you testing with only one connection or multiple?

**Comment by bobholt** [comment]
@BanzaiMan: Thanks for the response.

Yes, only running a single job locally. When I watch my Sauce dashboard while running the multiple jobs on Travis, I don't see it hitting the limit, but it's possible.

Is there a way to only open Sauce Connections for specific jobs? We only need to run Sauce on 2 of them.

**Comment by bobholt** [comment]
I can confirm that I'm not running up against any job limits in my Sauce Dashboard. Just ran a new build (https://travis-ci.org/bobholt/web-platform-tests/builds/221755382). There are at most 3 concurrent tunnels, and I never see one opened for 260.5. I see tunnels opened for all of the rest as I watch the Sauce dashboard.

**Comment by BanzaiMan** [comment]
[260.5](https://travis-ci.org/bobholt/web-platform-tests/jobs/221755387#L443) says it opened the tunnel.

**Comment by BanzaiMan** [comment]
> 13 Apr 12:59:15 - Remote tunnel VM has been terminated by user request. Shutting down.

What does this mean? Something is shutting down the tunnel, right?

**Comment by bobholt** [comment]
> 260.5 says it opened the tunnel.

260.5 may have opened the tunnel, but it closed so immediately that it didn't appear in the Sauce UI. I was replying in the context of your suggestion that there were too many connections to sauce, but that it appears that I have plenty to spare. 

My question still stands, however: Is there a way to only open Sauce Connections for specific jobs? This would be the easy way to make sure we aren't running out of connections, but I don't see any documentation for that.

If there is no way to do that with the Sauce Connect addon, then I will manually download, install, and run SauceConnect in the build script for only the jobs where we need that.

> What does this mean? Something is shutting down the tunnel, right?

Yes, that's what I'm trying to find out. I believed it to be an instability in Travis.

**Comment by BanzaiMan** [comment]
Where is the termination initiated? "Remote tunnel VM" means the VM running on Sauce Labs end, right?

**Comment by bobholt** [comment]
Yes, the VM on the Sauce Labs side. I believe there is a Travis error causing an early termination of the tunnel. The purpose of this issue is to find where that termination is initiated.

Is there a way to only open Sauce Connections for specific jobs so that we can eliminate that as a possibility?

**Comment by BanzaiMan** [comment]
Sauce Connect addon is triggered unconditionally. You can limit the number of jobs that can run simultaneously, however. https://docs.travis-ci.com/user/customizing-the-build/#Limiting-Concurrent-Builds

**Comment by bobholt** [comment]
That doesn't seem to help. Let me restate the facts as I see them:

1. The log output described above comes from Sauce only when the Sauce Connect process receives an interrupt (`ctrl-C`)-type event.
2. The Sauce Connect process is spawned by the Travis Sauce Connect addon.
3. The (my) build script is unaware of this process, has no mechanism to cause it to end, and does not attempt to do so.
4. Whatever is causing this interrupt:
  - is intermittent - happening about 50% of the time
  - only happens on Travis
  - happens even if there is only 1 job running on Travis and Sauce (e.g. when re-running a single job from a single build with no other builds happening simultaneously)

This all leads me to believe that there is a bug in the Travis Sauce Connect Addon _or_ an error somewhere else in the Travis system causing that process to _sometimes_ receive an interrupt signal.

**Comment by BanzaiMan** [comment]
Nothing happens in the Sauce Connect addon after the tunnel is open until your build gets past the `script` section. https://github.com/travis-ci/travis-build/blob/master/lib/travis/build/addons/sauce_connect.rb

**Comment by bobholt** [comment]
Something is causing that process to die. It's not intentional in that addon, but it is happening. Looking at your code also shows me that the SC process dying is why sometimes our builds fail with "No running Sauce Connect tunnel found" - because the PID you're looking for [here](https://github.com/travis-ci/travis-build/blob/f54f321f500b1f39a940d63333cbf3a1161c781c/lib/travis/build/addons/sauce_connect/templates/sauce_connect.sh#L80) has already been killed _for some other reason_.

I'm wondering if it's a VM memory issue. The tunnel closure _always_ happens while our build script is building a manifest, which is a memory-intensive process. In my job links above, the tunnel closure always happens after our `python manifest` script is invoked and before actually running any tests.

I see that a Sudo-enabled machine should have 7.5GB of memory, so I wouldn't expect that to be the issue, but still the occasionally-dying-process matches other similar out-of-memory bug reports (See #2101).

**Comment by bobholt** [comment]
Running that manifest script locally, I see that it uses 100% CPU and slowly more memory (up to 6.5%) until it completes. I see that CPU listing for sudo-enabled machines is "~2, bursted" - maybe the burst causes the SC process to die.

**Comment by BanzaiMan** [comment]
Hmmm. So, the kernel is killing `sc`? This is quite strange, as if this were the case, I would expect to see reports of this kind more often.

**Comment by bobholt** [comment]
The kernel may kill it, but in discussing it over here, it could also be that there is some sort of keepalive between the tunnel client and Sauce that is interrupted by the resource usage of the manifest build. It uses 100% CPU, over a gigabyte of memory, and (potentially) 100% of file I/O. While we need to optimize that script, part of that is by design in order to not "leave any performance on the table."

I seem to have been partially successful in not using the Travis Sauce Connect addon, but instead installing sauce connect and starting it as part of the build _after_ the manifest script has run. This gives us the added benefit of not opening a Sauce tunnel for the 5 other jobs that don't need it.

**Comment by magician03** [comment]
Hi,

I am facing this problem. Everything worked well and fine 2 days ago. All my SauceLabs Emulator/Browser tests ran well till 2 days ago. Now I am this `Failed: ETIMEDOUT connect ETIMEDOUT 162.222.75.45:80` issue.

I am pretty sure this is the problem related to Travis as everything is working well well and fine when I am connecting to sauce labs via my local machine. 

Here is one of the logs: [https://travis-ci.org/magician03/moodlemobile2/builds/240536982](https://travis-ci.org/magician03/moodlemobile2/builds/240536982)

What is the actual Issue ?, I am thinking that the connection to selenium address ondemand.saucelabs.com:80/wd/hub is being lost intermediately as Sauce is not receiving any commands after some time of the start of the test as Android Emulator on SauceLabs doesn't need SauceConnectProxy to be on ( the sauce-connect addon ). Is there any solution for this issue? or will it again work well after few like it did earlier?

**Comment by vmalloc** [comment]
I'm also hitting this issue. Interested to know if anyone found a workaround or a solution.

**Comment by bobholt** [comment]
My workaround was to download Sauce Connect, install, and start it as part of my Travis job. It still occasionally times out, but less often than the Travis addon. The Python code to do this in our setup is here: https://github.com/w3c/web-platform-tests/blob/master/tools/wptrunner/wptrunner/browsers/sauce.py#L138.
